message_signer.go

```diff
@@ -0,0 +1,254 @@
+	digestHeaderKey = "Digest"
+
+	signatureHeaderKeyIDKey     = "keyId"
+	signatureHeaderAlgorithmKey = "algorithm"
+	signatureHeaderHeadersKey   = "headers"
+	signatureHeaderSignatureKey = "signature"
+
+	signatureStringDelimiter = ": "
+	spaceSeparator           = " "
+	headerValueSeparator     = ", "
+	newline                  = "\n"
+	doubleQuote              = `"`
+	commaDelimiter           = ","
+	equalDelimiter           = "="
+)
+
+// TargetHeader is the header that the signature should be populated on.
+type TargetHeader string
+
+const (
+	Signature     TargetHeader = "Signature"
+	Authorization TargetHeader = "Authorization"
+
+	authorizationHeaderPrefix string = "Signature "
+)
+
+// valuePrefix returns a string that should be prefixed to the header value for the given target.
+func (s TargetHeader) valuePrefix() string {
+	switch s {
+	case Authorization:
+		return authorizationHeaderPrefix
+	case Signature:
+		return ""
+	default:
+		return ""
+	}
+}
+
+const (
+	// RequestTarget is the special case header that can be included in the signature string.
+	RequestTarget = "(request-target)"
+)
+
+// signatureStringSpecialCase specifies format functions for designated headers.
+// More information can be found in section 2.3 [here](https://www.ietf.org/archive/id/draft-cavage-http-signatures-12.txt).
+func fetchSignatureStringSpecialCase(s string) func(r *http.Request, b *bytes.Buffer) error {
+	switch s {
+	case RequestTarget:
+		return addRequestTarget
+	default:
+		return nil
+	}
+}
+
+// MessageSigner is used to sign HTTP messages.
+type MessageSigner struct {
+	// algo is the chosen crypto algorithm.
+	algo *HashingAlgorithm
+	// signer will be used to sign the message.
+	signer Signer
+	// publicKeyID is the ID that can be used to look up the public key on the receiving server.
+	publicKeyID string
+	// targetHeader defines which header to populate on the request with the signature.
+	targetHeader TargetHeader
+}
+
+// NewMessageSigner checks that the given algorithm is valid and returns a new MessageSigner.
+func NewMessageSigner(algo crypto.Hash, signer Signer, publicKeyID string, targetHeader TargetHeader) (*MessageSigner, error) {
+	cryptoHash, err := fetchCryptoHash(algo)
+	if err != nil {
+		return nil, NewInitialisationError("could not retrieve crypto hash", err)
+	}
+
+	return &MessageSigner{
+		algo:         cryptoHash,
+		publicKeyID:  publicKeyID,
+		signer:       signer,
+		targetHeader: targetHeader,
+	}, nil
+}
+
+// SignRequest method signs provided http.Request
+// signatureHeaders is a list of header names that specifies which of them (together with their values) will be signed. At least one is required.
+func (ms *MessageSigner) SignRequest(req *http.Request, signatureHeaders []string) (*http.Request, error) {
+	if len(signatureHeaders) == 0 {
+		return nil, NewValidationError("no signature headers set, at least one should be specified")
+	}
+
+	body, err := readBody(req)
+	if err == nil && len(body) != 0 {
```

Comment:
I think the error is ignored here if not empty, shouldn't we return the error in case reading the body fails?

Reply:
good catch